### ISSUE TITLE
Introduce ModifierChain.copyAndExclude()?

### ISSUE THREAD

**Issue by makortel** [cms-sw/cmssw#15496]
Especially now that the era definitions are scattered in separate files (#15480), I think a `ModifierChain.copyAndExclude()` or similar would make it easier keep certain sets of eras synchronized. As an example `Run2_2017_trackingPhase1PU70` and `Run2_2017_trackingRun2` are already out of synch with respect to `Run2_2017` (see my comment https://github.com/cms-sw/cmssw/pull/15469#issuecomment-239787517, and PR fixing the issue #15495).

@Dr15Jones

**Comment by Dr15Jones** [comment]
assign core

**Comment by Dr15Jones** [comment]
I don't see how that could be made to work. If you had

---foo_era.py----------
```python
foo = cms.Modifier()
```

---bar_era.py---------
```python
from foo_era import foo as _foo
import something_using_foo

bar = _foo.copyAndExclude(something_using_foo.bleck)
```

So now you've inverted the dependency. Instead of having `bar_era.py` imported in many files, you have `bar_era.py` importing many files and I think we are back to the problem we had with `customise` functions.

**Comment by Dr15Jones** [comment]
Could the problem with `Run2_2017` and the `Run2_2017_*` be solved by making a `Run2_2017_core` that they all share?

**Comment by Dr15Jones** [comment]
So I see that you were referring to `ModifierChain` not `Modifier` (I blame insufficient caffeine this morning :) ). However, how do you know that if an additional `Modifier` is added to `Run2_2017` that you'd want it in `Run2_2017_*`? You'd still have to review each use of it and decide if they need to be modified.

**Comment by makortel** [comment]
I agree that in general one would need to review all cases. In this particular case the two other `Run2_2017_*` `ModifierChain`s are (logically) defined as `Run2_2017` but with `trackingPhase1` 
`Modifier` replaced with `trackingPhase1PU70` or nothing, so whatever is added to `Run2_2017` should be propagated to the other two. Ok, it would be good if one would nevertheless review the other two aswell (so we should pay more attention to modifications of existing `ModifierChain`s), but with "copyAndExclude" the definitions of the other two would still be shorter.

I think `Run2_2017_core` would actually solve the "problem", thanks for the suggestion. I can give it a try after #15495 gets merged and we'll see if somebody disagrees.

**Comment by makortel** [comment]
The suggestion finally materialized in #15994.

**Comment by makortel** [comment]
The `Run2_2017_core` approach seems to be good enough so I close this issue.

**Comment by makortel** [comment]
Reopening as #17605 presents a use case where the `Run2_2017_core` approach does not work well.

@kpedro88

**Comment by davidlange6** [comment]
sometimes a redesign of the hierarchy of objects can be needed due to changes in requirements..

**Comment by Dr15Jones** [comment]
@davidlange6 are you advocating making this change or for reworking the Era hierarchy?

**Comment by davidlange6** [comment]
maybe we should discuss starting from an example and decide if we want to support that example or not?

**Comment by davidlange6** [comment]
ok, I think I understand the example 

There are N (= #of detectors + #of POGs +++) potential groups that want variants of the “default” for each CMS running scenario. In this case “default” is the 2017 configuration baseline.

Matti was the first consumer, so he pushed his tracking configuration out of the “core era” and into an upper level. Then tracking could add its variants 

Now… how does the second consumer do the same thing? He has to take his config out of the “core era” and into an upper level. But this means the tracking has to change all of its variants to include this change.

Then, what should happen when the default of the “second consumer” changes? Then they have to look for all other variants of everyone and change them

So indeed, a copyAndExclude would allow these variants to involve independently.

**Comment by kpedro88** [comment]
@davidlange6 exactly

**Comment by Dr15Jones** [comment]
It seems to me that `copyAndExclude` doesn't decrease the `ModifierChains` it just creates implicit ones. That is fine if it is used in one and only one place, but the moment the exact same arguments have to be passed to another `copyAndExclude` I'd think you'd have a maintenance burden.

I would think having explicit names would make it easier to follow the era connections rather than having to also interpret those connections in addition to understanding all the applied `copyAndExclude`.

**Comment by makortel** [comment]
If I understood Kevin's intention in #17605 correctly, we would need something along (moving `run2_GEM_2017` to current `Run2_2017_core` where it should have been in the first place):
```
# Intermediate ModifierChains
Run2_2017_core = [Run2_2016, phase1Pixel, run2_HF_2017, run2_HCAL_2017, run2_GEM_2017, run2_HE_2017]
Run2_2017_core_HEPlan1 = [Run2_2017_core, run2_HEPlan1_2017]
Run2_2017_core_noHEPlan1_tracking = [Run2_2017_core, trackingPhase1QuadProp]

# Final ModifierChains
Run2_2017 = [Run2_2017_core_HEPlan1, trackingPhase1QuadProp]
Run2_2017_trackingPhase1CA = [Run2_2017_core_HEPlan1, trackingPhase1]
Run2_2017_trackingRun2 = [Run2_2017_core_HEPlan1]

Run2_2018 = [Run2_2017_core_noHEPlan1_tracking]
```
(details of HCAL sub-eras can be incorrect, but hopefully close enough to make the point)

Same exercise with copyAndExclude
```
Run2_2017 = [Run2_2016, phase1Pixel, run2_HF_2017, run2_HCAL_2017, run2_GEM_2017, run2_HE_2017, run2_HEPlan1_2017, trackingPhase1QuadProp]
Run2_2017_trackingPhase1CA = Run2_2017.copyAndExclude([trackingPhase1QuadProp])+[trackingPhase1]
Run2_2017_trackingRun2 = Run2_2017.copyAndExclude([trackingPhase1QuadProp])

Run2_2018 = Run2_2017.copyAndExclude([run2_HEPlan1_2017])
```

There is some amount of repeated information (2017 default tracking sub-era being `trackingPhase1QuadProp`) in both that seems to be difficult to avoid. To me the latter would be clearer to read, because it would consist from 4 files instead of 7 (since currently each `ModifierChain` is in its own file), and the mental graph of the `ModifierChains` is simpler.

In case a third party (or more) would like to join the game, I suspect the first option scales multiplicatively, while the latter would be additive.

In either case, we (I) should probably put more comments in the `ModifierChain` files to instruct more explicitly "what other files need to be checked/changed if this file is changed".

**Comment by makortel** [comment]
Let me add after a bit more thought that in my latter example above, I'd likely write
```
Run2_2017_trackingRun2 = Run2_2017.copyAndExclude([trackingPhase1QuadProp])
Run2_2017_trackingPhase1CA = [Run2_2017_trackingRun2, trackingPhase1]
```
so I'd be fine if we impose a rule that a `Modifier` can be excluded from the same `ModifierChain` only once (and if more times is needed, better make a named intermediate `ModifierChain`).

**Comment by kpedro88** [comment]
@makortel I definitely agree about multiplicative vs additive scaling, that's why I was hesitant to start refactoring things again.

**Comment by Dr15Jones** [comment]
#17612 
I'm still not convinced this is the best idea but it takes less effort to make it than to continue discussing it :).

**Comment by makortel** [comment]
Closing as the feature was implemented in #17612 and taken in use in #17644.
